cunumeric/coverage.py

```diff
@@ -96,18 +96,28 @@ def wrapper(*args: Any, **kwargs: Any) -> Any:
             location = find_last_user_frames(
                 not runtime.args.report_dump_callstack
             )
+            if set_provenance := (runtime.legate_context.provenance is None):
+                runtime.legate_context.set_provenance(find_last_user_frames())
             runtime.record_api_call(
                 name=name,
                 location=location,
                 implemented=True,
             )
-            return func(*args, **kwargs)
+            result = func(*args, **kwargs)
+            if set_provenance:
+                runtime.legate_context.reset_provenance()
+            return result
 
     else:
 
         @wraps(func)
         def wrapper(*args: Any, **kwargs: Any) -> Any:
-            return func(*args, **kwargs)
+            if set_provenance := (runtime.legate_context.provenance is None):
```

Comment:
This is to activate provenance tracking only for top-level invocations; some cuNumeric operators internally invoke other cuNumeric operators and because all of them share the same provenance, we need to set the provenance only once for the very first invocation made by the user program.

Reply:
This seems fine but a context manager could be slightly nicer to automatically handle the call the `reset_provenance` when necessary. BTW is there any reason not to just call `reset_provenance` unconditionally? If `legate_context.provenance` was set, we need to reset it, but if `legate_context.provenance` was not set, then presumably resetting it is a no-op?

Reply:
> This seems fine but a context manager could be slightly nicer to automatically handle the call the reset_provenance when necessary.

I guess I could do something like this:
```
func = runtime.legate_context.track_provenance(func, callback=find_last_user_frames)
```
and internally it calls the callback only when the provenance hasn't been set. Let me push changes for this.

>  BTW is there any reason not to just call reset_provenance unconditionally? If legate_context.provenance was set, we need to reset it, but if legate_context.provenance was not set, then presumably resetting it is a no-op?

No you don't always want to reset the provenance if you're already within a cunumeric API, otherwise, subsequent cuNumeric calls will lose the provenance. For example, in case you have a cuNumeric op A that internally calls cuNumeric ops B and C, unconditional resetting will reset the provenance as soon as it returns from B, which might not launch anything (e.g., if it was `ndarray.__init__`), thereby not tracking provenance for C.

Reply:
I meant something like this just to wrap up the code that is repeated here a few places::
```
@contextmanager
def track_provenance():
    if set_provenance := (runtime.legate_context.provenance is None):
        runtime.legate_context.set_provenance(find_last_user_frames())
    yield set_provenance
    if set_provenance:
                runtime.legate_context.reset_provenance()
```
Then
```
with track_provenance():
    return func()
```


> No you don't always want to reset the provenance if you're already within a cunumeric API, otherwise, subsequent cuNumeric calls will lose the provenance.

got it

Reply:
alright, I added a decorator for provenance tracking and it supports both top-level only tracking and nested tracking: https://github.com/nv-legate/cunumeric/pull/596/commits/8448aec8bcb42595befa2c929facf7da2f86e5ef please also take a look at the companion PR: https://github.com/nv-legate/legate.core/pull/370